kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy;
+
+import java.util.Map;
+import java.util.Optional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.config.Configuration;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * The proxy can operate in one of two modes:
+ * <ol>
+ *     <li>
+ *         {@linkplain #PRODUCTION} mode (default). Development features are disallowed.
+ *     </li>
+ *     <li>
+ *         {@linkplain #DEVELOPMENT} mode. Development features are allowed
+ *     </li>
+ * </ol>
+ */
+public enum ProxyMode {
+    PRODUCTION {
+        @Override
+        public Configuration validate(@NonNull Configuration config) {
+            Optional<Map<String, Object>> development = config.development();
+            if (development.isPresent() && !development.get().isEmpty()) {
```

Comment:
I can't help but think insisting that all development mode options are expressed at the top level seems overly restrictive. 

What if a filter wants to use it wants it own development configuration? I'm wondering if we actually want to an annotation (`@DevelopmentMode`) that could be applied anywhere on the config model.   The proxy, in production mode, would then walk the configuration tree and check for any @DevelopmentMode annotated fields that were non-null (or Optional not empty).

We could also use the same configuration tree walker to report on use of any deprecated configuration items in a standard way.

WDYAT? If we decide to merge #1648 as it stands (to break the log jam) that's fine too..

Reply:
Keen to merge this as-is and get the branches in. 

Definitely makes sense to have it available anywhere in the tree so you don't have to cobble together nested structures in this `development` object. But I think we should add it later.

Reply:
I think the way we should do this is:

* Generate the proxy config from schema
* Allow unknown properties on all objects in the schema
* Run a visitor over the configuration instance to check whether there are any unknown properties
* In normal mode we fail if we there are unknown properties, but in normal mode we allow it with a warning logged about the configuration using non-supported properties which have no compatibility guarantees and may disappear without notice in the future.

There are a few choices to be made in exactly how we want that to work. 

I'm not sure I'd call this development mode. That makes it sound kinda _interesting_ to developers, which is the opposite of what we want. It's also ambiguous about what kind of developer (filter author, someone working on kroxy itself, but also our users might also identify with developer). So to me this is more of a `-Dunlock_system_test_only_configs`. 

I guess I'm OK with this merging this PR to unblock, but can we add some wording about "supported properties which have no compatibility guarantees" to the warning we log?

Reply:
I share @k-wall's concern about the usefulness of only having development options at the root however its easy to relax restrictions later when we have a use case or requirement for it rather than just guessing it might be useful.

Reply:
Happy to make it more specific.

The `unlock_*` pattern would point to a slightly different implementation, feature based. So where I'm passing around a mode it would be more like a `Set<Feature>`.

Reply:
> Happy to make it more specific.
> 
> The `unlock_*` pattern would point to a slightly different implementation, feature based. So where I'm passing around a mode it would be more like a `Set<Feature>`.

Does that just suggest we should be using a feature flag library? Such as https://openfeature.dev/

Reply:
Personally I've always struggled to get away from the feeling that open feature is just over engineering. It's certainly too soon to decide something like that when the problem we're trying to solve right now is one where we have single use case for needing to configure something like this in very specific circumstances.